dependencies-in-practice.Rmd

```diff
@@ -0,0 +1,730 @@
+    #' @importFrom glue glue_collapse
+    ## usethis namespace: end
+    NULL
+    ```
+
+Recall that `devtools::document()` processes your roxygen comments (@sec-dependencies-NAMESPACE-workflow), which writes help topics to `man/*.Rd` and, relevant to our current goal, generates the `NAMESPACE` file.
+If you use `use_import_from()`, it does this for you and also calls `load_all()`, making the newly imported function available in your current session.
+
+The roxygen tag above cause this directive to appear in the `NAMESPACE` file:
+
+    importFrom(glue, glue_collapse)
+
+Now you can use the imported function directly in your code:
+
+```{r}
+#| eval: false
+somefunction <- function(...) {
+  ...
+  x <- glue_collapse(...)
+  ...
+}
+```
+
+Sometimes you make such heavy use of so many functions from another package that you want to import its entire namespace.
+This should be relatively rare.
+In the tidyverse, the package we most commonly treat this way is rlang, which functions almost like a base package for us.
+
+Here is the roxygen tag that imports all of rlang.
+This should appear somewhere in `R/*.R`, such as the dedicated space described above for collecting all of your namespace import tags.
+
+```{r}
+#| eval: false
+#' @import rlang
+```
+
+After calling `devtools::document()`, this roxygen tag causes this directive to appear in the `NAMESPACE` file:
+
+    import(rlang)
+
+This is the least recommended solution because it can make your code harder to read (you can't tell where a function is coming from), and if you `@import` many packages, it increases the chance of function name conflicts.
+Save this for very special situations.
+
+#### How to *not* use a package in Imports
+
+Sometimes you have a package listed in `Imports`, but you don't actually use it inside your package or, at least, R doesn't think you use it.
+That leads to a `NOTE` from `R CMD check`:
+
+    * checking dependencies in R code ... NOTE
+    Namespace in Imports field not imported from: ‘aaapkg’
+      All declared Imports should be used.
+
+This can happen if your only use of aaapkg is a call like `aaapkg::aaa_fun()` in the default for a function argument in your package.
+This also comes up if you want to list an indirect dependency in `Imports`, so you can state a minimum version for it.
+The tidyverse meta-package has this problem on a large scale, since it exists mostly to install a bundle of packages at specific versions.
+
+How can you get rid of this `NOTE`?
+
+Our recommendation is to put a namespace-qualified reference (not a call) to an object in aaapkg in some file below `R/`, such as a `.R` file associated with package-wide setup:
+
+```{r}
+#| eval: false
+ignore_unused_imports <- function() aaapkg::aaa_fun
+```
+
+You don't need to call `ignore_unused_imports()` anywhere.
+You shouldn't export it.
+You don't have to actually exercise `aaapkg::aaa_fun()`.
+What's important is to access something in aaapkg's namespace with `::`.
```

Comment:
I didn't add this because the wording seems off. We aren't, in fact, importing aaapkg. We aren't even loading it. And this same code in a signature doesn't silence the NOTE.

If you think this point is important, let's find better wording.

Reply:
I meant importing in the NAMESPACE sense 😬